File: common/src/main/kotlin/generations/gg/rarecandylib/common/client/model/GenerationsModelLoader.java
package generations.gg.rarecandylib.common.client.model;

import gg.generations.rarecandy.pokeutils.PixelAsset;
import gg.generations.rarecandy.renderer.components.MeshObject;
import gg.generations.rarecandy.renderer.components.MultiRenderObject;
import gg.generations.rarecandy.renderer.loading.ModelLoader;
import gg.generations.rarecandy.renderer.model.GLModel;

import java.io.InputStream;
import java.util.ArrayList;
import java.util.function.Supplier;

public class GenerationsModelLoader extends ModelLoader {
    public GenerationsModelLoader(int numThreads) {
        super(numThreads);
    }
    public MultiRenderObject<MeshObject> compiledModelMethod(CompiledModel model, InputStream stream, Supplier<MeshObject> supplier, String name, boolean requiresVariantTexture) {
        return createObject(
                () -> new PixelAsset(stream, name),
                (gltfModel, animResources, textures, config, object) -> {
                    var glCalls = new ArrayList<Runnable>();
                    try {
//                        if(GenerationsCore.CONFIG.client.useVanilla) {
//                            VanillaModelLoader.processModel(object, gltfModel, animResources, textures, config, glCalls, supplier);
//                        } else {
                            processModel(object, gltfModel, animResources, textures, config, glCalls, supplier, GLModel::new);
//                        }
                    } catch (Exception e) {
                        System.out.println("Oh no! Model : " + name + " didn't properly load!");
                        e.printStackTrace();
                    }
                    return glCalls;
                },
                object -> {
                    if(object.scale == 0f) object.scale = 1.0f;

//                    if(GenerationsCore.CONFIG.client.logModelLoading) GenerationsCore.LOGGER.info("Done Loading: " + name);
                }
        );
    }
}

File: common/src/main/kotlin/generations/gg/rarecandylib/common/client/render/rarecandy/loading/GenerationsModelLoader.kt
package generations.gg.rarecandylib.common.client.render.rarecandy.loading

import generations.gg.rarecandylib.common.client.render.rarecandy.CobblemonInstance
import generations.gg.rarecandylib.common.client.render.rarecandy.CompiledModel
import gg.generations.rarecandy.pokeutils.PixelAsset
import gg.generations.rarecandy.renderer.components.MeshObject
import gg.generations.rarecandy.renderer.components.MultiRenderObject
import gg.generations.rarecandy.renderer.loading.ModelLoader
import gg.generations.rarecandy.renderer.model.GLModel
import org.joml.Matrix4f
import java.io.InputStream
import java.util.function.Supplier

class GenerationsModelLoader(numThreads: Int) : ModelLoader(numThreads) {
    fun compiledModelMethod(
        model: CompiledModel,
        stream: InputStream,
        supplier: Supplier<MeshObject>,
        name: String?,
        requiresVariantTexture: Boolean
    ): MultiRenderObject<MeshObject> {
        return createObject<MeshObject>(
            { PixelAsset(stream, name) },
            { gltfModel, animResources, textures, config, `object` ->
                val glCalls = ArrayList<Runnable>()
                try {
                    processModel<MeshObject?, MultiRenderObject<MeshObject?>?>(
                        `object`,
                        gltfModel,
                        animResources,
                        textures,
                        config,
                        glCalls,
                        supplier
                    ) { vertexBuffer, indexBuffer, glCalls, indexSize, gltType, attributes ->
                        GLModel(
                            vertexBuffer,
                            indexBuffer,
                            glCalls,
                            indexSize,
                            gltType,
                            attributes
                        )
                    }
                } catch (e: Exception) {
                    println("Oh no! Model : " + name + " didn't properly load!")
                    e.printStackTrace()
                }
                glCalls
            },
            { `object`: MultiRenderObject<MeshObject?>? ->
                model.guiInstance = CobblemonInstance(Matrix4f(), Matrix4f(), null)
                model.guiInstance!!.link(`object`)
                if (`object`!!.scale == 0f) `object`.scale = 1.0f
            }
        )
    }
}